File: src/main/java/dev/hoot/api/packets/ItemPackets.java
package dev.hoot.api.packets;

import dev.hoot.api.game.Game;
import dev.hoot.api.widgets.Widgets;
import meteor.api.packets.ClientPackets;
import net.runelite.api.Client;
import net.runelite.api.InventoryID;
import net.runelite.api.Item;
import net.runelite.api.packets.ClientPacket;
import net.runelite.api.packets.PacketBufferNode;
import net.runelite.api.widgets.WidgetInfo;
import net.runelite.packets.ObfuscatedClientPacket;

import java.util.List;

public class ItemPackets
{
	public static void itemAction(Item item, String action)
	{

		List<String> actions = List.of(Game.getClient().getItemDefinition(item.getId()).getInventoryActions());
		int index = actions.indexOf(action);
		System.out.println(index);
		switch (index) {
			case 0 -> itemFirstOption(item);
			case 1 -> itemSecondOption(item);
			case 2 -> itemThirdOption(item);
			case 3 -> itemFourthOption(item);
			case 4 -> itemFifthOption(item);
			default -> WidgetPackets.widgetAction(Widgets.fromId(item.getWidgetId()), action);
		}
	}

	public static void itemFirstOption(Item item)
	{
		queueItemAction1Packet(item.getWidgetId(), item.getId(), item.getSlot());
	}

	public static void itemSecondOption(Item item)
	{
		queueItemAction2Packet(item.getWidgetId(), item.getId(), item.getSlot());
	}

	public static void itemThirdOption(Item item)
	{
		queueItemAction3Packet(item.getWidgetId(), item.getId(), item.getSlot());
	}

	public static void itemFourthOption(Item item)
	{
		queueItemAction4Packet(item.getWidgetId(), item.getId(), item.getSlot());
	}

	public static void itemFifthOption(Item item)
	{
		queueItemAction5Packet(item.getWidgetId(), item.getId(), item.getSlot());
	}

	public static void useItemOnItem(Item item, Item item2)
	{
		if ((item.getType().getInventoryID() != InventoryID.INVENTORY) || (item2.getType().getInventoryID() != InventoryID.INVENTORY))
		{
			return;
		}

		queueItemOnItemPacket(item.getId(), item.getSlot(), item2.getId(), item2.getSlot());
	}

	public static void queueItemOnItemPacket(int sourceItemId, int sourceItemSlot, int itemId, int itemSlot)
	{
		createItemOnItem(sourceItemId, sourceItemSlot, itemId, itemSlot).send();
	}

	public static void queueSpellOnItemPacket(int itemId, int itemSlot, int spellWidgetId)
	{
		createSpellOnItem(itemId, itemSlot, spellWidgetId).send();
	}

	public static void queueItemAction1Packet(int itemWidgetId, int itemId, int itemSlot)
	{
		createFirstAction(itemWidgetId, itemId, itemSlot).send();
	}

	public static void queueItemAction2Packet(int itemWidgetId, int itemId, int itemSlot)
	{
		createSecondAction(itemWidgetId, itemId, itemSlot).send();;
	}

	public static void queueItemAction3Packet(int itemWidgetId, int itemId, int itemSlot)
	{
		createThirdAction(itemWidgetId, itemId, itemSlot).send();;
	}

	public static void queueItemAction4Packet(int itemWidgetId, int itemId, int itemSlot)
	{
		createFourthAction(itemWidgetId, itemId, itemSlot).send();;
	}

	public static void queueItemAction5Packet(int itemWidgetId, int itemId, int itemSlot)
	{
		createFifthAction(itemWidgetId, itemId, itemSlot).send();;
	}

	public static PacketBufferNode createFirstAction(int itemWidgetId, int itemId, int itemSlot)
	{
		ObfuscatedClientPacket packet = ClientPackets.INSTANCE.get("OPHELD1");
		return ClientPackets.INSTANCE.createItemPacket(packet, itemId, itemSlot, itemWidgetId);
	}

	public static PacketBufferNode createSecondAction(int itemWidgetId, int itemId, int itemSlot)
	{
		ObfuscatedClientPacket packet = ClientPackets.INSTANCE.get("OPHELD2");
		return ClientPackets.INSTANCE.createItemPacket(packet, itemId, itemSlot, itemWidgetId);
	}

	public static PacketBufferNode createThirdAction(int itemWidgetId, int itemId, int itemSlot)
	{
		ObfuscatedClientPacket packet = ClientPackets.INSTANCE.get("OPHELD3");
		return ClientPackets.INSTANCE.createItemPacket(packet, itemId, itemSlot, itemWidgetId);
	}

	public static PacketBufferNode createFourthAction(int itemWidgetId, int itemId, int itemSlot)
	{
		ObfuscatedClientPacket packet = ClientPackets.INSTANCE.get("OPHELD4");
		return ClientPackets.INSTANCE.createItemPacket(packet, itemId, itemSlot, itemWidgetId);
	}

	public static PacketBufferNode createFifthAction(int itemWidgetId, int itemId, int itemSlot)
	{
		ObfuscatedClientPacket packet = ClientPackets.INSTANCE.get("OPHELD5");
		return ClientPackets.INSTANCE.createItemPacket(packet, itemId, itemSlot, itemWidgetId);
	}

	public static PacketBufferNode createItemOnItem(int sourceItemId, int sourceItemSlot, int itemId, int itemSlot)
	{
		Client client = Game.getClient();
		ClientPacket clientPacket = Game.getClientPacket();
		PacketBufferNode packetBufferNode = Game.getClient().preparePacket(clientPacket.OPHELDU(), client.getPacketWriter().getIsaacCipher());
		packetBufferNode.getPacketBuffer().writeIntLE(WidgetInfo.INVENTORY.getId());
		packetBufferNode.getPacketBuffer().writeInt(WidgetInfo.INVENTORY.getId());
		packetBufferNode.getPacketBuffer().writeShortLE(sourceItemSlot);
		packetBufferNode.getPacketBuffer().writeShortLE(sourceItemId);
		packetBufferNode.getPacketBuffer().writeShortAdd(itemSlot);
		packetBufferNode.getPacketBuffer().writeShortAddLE(itemId);
		return packetBufferNode;
	}

	public static PacketBufferNode createSpellOnItem(int itemId, int itemSlot, int spellWidgetId)
	{
		Client client = Game.getClient();
		ClientPacket clientPacket = Game.getClientPacket();
		PacketBufferNode packetBufferNode = Game.getClient().preparePacket(clientPacket.OPHELDT(), client.getPacketWriter().getIsaacCipher());
		packetBufferNode.getPacketBuffer().writeShortAdd(itemId);
		packetBufferNode.getPacketBuffer().writeShort(-1);
		packetBufferNode.getPacketBuffer().writeInt(spellWidgetId);
		packetBufferNode.getPacketBuffer().writeIntIME(WidgetInfo.INVENTORY.getId());
		packetBufferNode.getPacketBuffer().writeShortLE(itemSlot);
		return packetBufferNode;
	}
}

File: src/main/kotlin/meteor/api/packets/ItemPackets.kt
package meteor.api.packets

import meteor.api.widgets.Widgets
import net.runelite.api.InventoryID
import net.runelite.api.Item
import net.runelite.api.widgets.WidgetInfo

object ItemPackets {
    fun itemAction(item: Item, action: String) {
        val actions = item.actions
        val index = actions.indexOf(action)
        when (index) {
            0 -> itemFirstOption(item)
            1 -> itemSecondOption(item)
            2 -> itemThirdOption(item)
            3 -> itemFourthOption(item)
            4 -> itemFifthOption(item)
            else -> WidgetPackets.widgetAction(Widgets.fromId(item.widgetId)!!, action)
        }
    }

    fun itemFirstOption(item: Item) {
        queueItemAction1Packet(item.widgetId, item.id, item.slot)
    }

    fun itemSecondOption(item: Item) {
        queueItemAction2Packet(item.widgetId, item.id, item.slot)
    }

    fun itemThirdOption(item: Item) {
        queueItemAction3Packet(item.widgetId, item.id, item.slot)
    }

    fun itemFourthOption(item: Item) {
        queueItemAction4Packet(item.widgetId, item.id, item.slot)
    }

    fun itemFifthOption(item: Item) {
        queueItemAction5Packet(item.widgetId, item.id, item.slot)
    }

    fun useItemOnItem(item: Item, item2: Item) {
        if (item.type.inventoryID != InventoryID.INVENTORY || item2.type.inventoryID != InventoryID.INVENTORY) {
            return
        }
        queueItemOnItemPacket(item.id, item.slot, item2.id, item2.slot)
    }

    fun queueItemOnItemPacket(sourceItemId: Int, sourceItemSlot: Int, itemId: Int, itemSlot: Int) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode =
            client.preparePacket(clientPacket.OPHELDU(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeIntLE(WidgetInfo.INVENTORY.id)
        packetBufferNode.packetBuffer.writeInt(WidgetInfo.INVENTORY.id)
        packetBufferNode.packetBuffer.writeShortLE(sourceItemSlot)
        packetBufferNode.packetBuffer.writeShortLE(sourceItemId)
        packetBufferNode.packetBuffer.writeShortAdd(itemSlot)
        packetBufferNode.packetBuffer.writeShortAddLE(itemId)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueSpellOnItemPacket(itemId: Int, itemSlot: Int, spellWidgetId: Int) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode =
            client.preparePacket(clientPacket.OPHELDT(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShortAdd(itemId)
        packetBufferNode.packetBuffer.writeShort(-1)
        packetBufferNode.packetBuffer.writeInt(spellWidgetId)
        packetBufferNode.packetBuffer.writeIntIME(WidgetInfo.INVENTORY.id)
        packetBufferNode.packetBuffer.writeShortLE(itemSlot)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueItemAction1Packet(itemWidgetId: Int, itemId: Int, itemSlot: Int) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode =
            client.preparePacket(clientPacket.OPHELD1(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShortLE(itemSlot)
        packetBufferNode.packetBuffer.writeShortLE(itemId)
        packetBufferNode.packetBuffer.writeInt(itemWidgetId)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueItemAction2Packet(itemWidgetId: Int, itemId: Int, itemSlot: Int) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode =
            client.preparePacket(clientPacket.OPHELD2(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShort(itemSlot)
        packetBufferNode.packetBuffer.writeIntME(itemWidgetId)
        packetBufferNode.packetBuffer.writeShort(itemId)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueItemAction3Packet(itemWidgetId: Int, itemId: Int, itemSlot: Int) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode =
            client.preparePacket(clientPacket.OPHELD3(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeIntIME(itemWidgetId)
        packetBufferNode.packetBuffer.writeShortAdd(itemSlot)
        packetBufferNode.packetBuffer.writeShortAddLE(itemId)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueItemAction4Packet(itemWidgetId: Int, itemId: Int, itemSlot: Int) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode =
            client.preparePacket(clientPacket.OPHELD4(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeInt(itemWidgetId)
        packetBufferNode.packetBuffer.writeShortLE(itemId)
        packetBufferNode.packetBuffer.writeShortAdd(itemSlot)
        client.packetWriter.queuePacket(packetBufferNode)
    }

    fun queueItemAction5Packet(itemWidgetId: Int, itemId: Int, itemSlot: Int) {
        val client = Main.client
        val clientPacket = Main.clientPacket
        val packetBufferNode =
            client.preparePacket(clientPacket.OPHELD5(), client.packetWriter.isaacCipher)
        packetBufferNode.packetBuffer.writeShortAdd(itemSlot)
        packetBufferNode.packetBuffer.writeShortLE(itemId)
        packetBufferNode.packetBuffer.writeInt(itemWidgetId)
        client.packetWriter.queuePacket(packetBufferNode)
    }
}